src/bstr.rs

```diff
@@ -0,0 +1,36 @@
+use crate::WMIError;
+use winapi::{
+    shared::wtypes::BSTR,
+    shared::wtypesbase::OLECHAR,
+    um::oleauto::*,
+};
+use std::convert::TryFrom;
+use std::ptr::{null, NonNull};
+
+/// A non-null [BSTR]
+///
+/// [BSTR]:         https://docs.microsoft.com/en-us/previous-versions/windows/desktop/automat/bstr
+pub(crate) struct BStr(NonNull<OLECHAR>);
+
+impl std::ops::Drop for BStr {
+    fn drop(&mut self) {
+        unsafe { SysFreeString(self.as_bstr()) }
+    }
+}
+
+impl BStr {
+    pub fn from_str(s: &str) -> Result<Self, WMIError> {
+        let len = s.encode_utf16().count();
+        let len32 = u32::try_from(len).map_err(|_| WMIError::ConvertLengthError(len as _))?;
+
+        let bstr = BStr(NonNull::new(unsafe { SysAllocStringLen(null(), len32) }).ok_or(WMIError::ConvertAllocateError)?);
```

Comment:
Can we do it like `oaidl`?

```
let s_utf16 = s.encode_utf16();
let len = s_utf16 .count();
... unsafe { SysAllocStringLen(s_utf16.as_ptr(), len32) } ...
```

(This also eliminates the second call to `encode_utf16()`)

Reply:
oaidl has an easier time of things as that string is already UTF16, whereas `s` is UTF8

I *could* do:
```rust
let s = s.encode_utf16().collect::<Vec<_>>();
let len = s.len();
... unsafe { SysAllocStringLen(s.as_ptr(), len32) } ...
```
While this would technically get rid of the second call to `encode_utf16`, this would:
1. Add one or more extra allocations (for the temporary Vec, which is immediately discarded)
2. Still iterate over the data twice or more (once to copy into the Vec, ?? copies to larger allocations when Vec is growing, and once to copy from the Vec into the BSTR inside SysAllocStringLen)

std::str::[EncodeUTF16](https://doc.rust-lang.org/std/str/struct.EncodeUtf16.html) doesn't implement [ExactSizeIterator](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html) / len, as the number of u16 code units a given UTF8 string converts into varies depending on the exact contents.

That said, collecting a Vec would simplify this code, so perhaps it's still preferrable?

Reply:
Ah I forgot it returned an iterator. Keep the original version!